src/ripple/app/ledger/impl/TimeoutCounter.h

```diff
@@ -0,0 +1,140 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_LEDGER_TIMEOUTCOUNTER_H_INCLUDED
+#define RIPPLE_APP_LEDGER_TIMEOUTCOUNTER_H_INCLUDED
+
+#include <ripple/app/main/Application.h>
+#include <ripple/beast/clock/abstract_clock.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/core/Job.h>
+#include <boost/asio/basic_waitable_timer.hpp>
+#include <mutex>
+
+namespace ripple {
+
+/**
+    This class is an "active" object. It maintains its own timer
+    and dispatches work to a job queue. Implementations derive
+    from this class and override the abstract hook functions in
+    the base.
+
+    This class implements an asynchronous loop:
+
+    1. The entry point is `setTimer`.
+
+    2. After `mTimerInterval`, `queueJob` is called, which schedules a job to
+       call `invokeOnTimer` (or loops back to setTimer if there are too many
+       concurrent jobs).
+
+    3. The job queue calls `invokeOnTimer` which either breaks the loop if
+       `isDone` or calls `onTimer`.
+
+    4. `onTimer` is the only "real" virtual method in this class. It is the
+       callback for when the timeout expires. Generally, its only responsibility
+       is to set `mFailed = true`. However, if it wants to implement a policy of
+       retries, then it has a chance to just increment a count of expired
+       timeouts.
+
+    5. Once `onTimer` returns, if the object is still not `isDone`, then
+       `invokeOnTimer` sets another timeout by looping back to setTimer.
+
+   This loop executes concurrently with another asynchronous sequence,
+   implemented by the subtype, that is trying to make progress and eventually
+   set `mComplete = true`. While it is making progress but not complete, it
+   should set `mProgress = true`, which is passed to onTimer so it can decide
+   whether to postpone failure and reset the timeout. However, if it can
+   complete all its work in one synchronous step (while it holds the lock), then
+   it can ignore `mProgress`.
+
+   This class is split from the PeerSet class.  // TODO remove after PR
+*/
+class TimeoutCounter
+{
+protected:
+    using ScopedLockType = std::unique_lock<std::recursive_mutex>;
+
+    struct QueueJobParameter
+    {
+        JobType jobType;
+        std::string jobName;
+        std::optional<std::uint32_t> jobLimit;
+    };
+
+    TimeoutCounter(
+        Application& app,
+        uint256 const& targetHash,
+        std::chrono::milliseconds timeoutInterval,
+        QueueJobParameter&& jobParameter,
+        beast::Journal journal);
+
+    virtual ~TimeoutCounter() = default;
+
+    /** Schedule a call to queueJob() after mTimerInterval. */
+    void
+    setTimer();
+
+    /** Queue a job to call invokeOnTimer(). */
+    virtual void
+    queueJob();
```

Comment:
This can now be private and non-virtual. It could even be a static function in `TimeoutCounter.cpp`, or just inlined (it's only called in one place, `setTimer()`).